src/Components/Components/src/Routing/QueryParameterValueSupplier.cs

```diff
@@ -0,0 +1,186 @@
+            }
+        }
+
+        private static SortedDictionary<ReadOnlyMemory<char>, List<QueryParameterDestination>>? BuildQueryParameterMappings(Type componentType, out int destinationsCount)
+        {
+            var candidateProperties = MemberAssignment.GetPropertiesIncludingInherited(componentType, ComponentProperties.BindablePropertyFlags);
+            SortedDictionary<ReadOnlyMemory<char>, List<QueryParameterDestination>>? result = null;
+            destinationsCount = 0;
+
+            foreach (var propertyInfo in candidateProperties)
+            {
+                if (!propertyInfo.IsDefined(typeof(ParameterAttribute)))
+                {
+                    continue;
+                }
+
+                var fromQueryAttribute = propertyInfo.GetCustomAttribute<SupplyParameterFromQueryAttribute>();
+                if (fromQueryAttribute is not null)
+                {
+                    // Found a parameter that's assignable from querystring
+                    var componentParameterName = propertyInfo.Name;
+                    var queryParameterName = (string.IsNullOrEmpty(fromQueryAttribute.Name)
+                        ? componentParameterName
+                        : fromQueryAttribute.Name).AsMemory();
+
+                    // If it's an array type, capture that info and prepare to parse the element type
+                    Type effectiveType = propertyInfo.PropertyType;
+                    var isArray = false;
+                    if (effectiveType.IsArray)
+                    {
+                        isArray = true;
+                        effectiveType = effectiveType.GetElementType()!;
+                    }
+
+                    if (!UrlValueConstraint.TryGetByTargetType(effectiveType, out var parser))
+                    {
+                        throw new NotSupportedException($"Querystring values cannot be parsed as type '{propertyInfo.PropertyType}'.");
+                    }
+
+                    // Append a destination list entry for this component parameter name
+                    result ??= new(QueryParameterNameComparer.Instance);
+                    if (!result.TryGetValue(queryParameterName, out var list))
+                    {
+                        result[queryParameterName] = list = new();
+                    }
+                    list.Add(new QueryParameterDestination(componentParameterName, parser, isArray, destinationsCount++));
+                }
+            }
+
+            return result;
+        }
+
+        private readonly struct QueryParameterDestination
+        {
+            public readonly string ComponentParameterName;
+            public readonly UrlValueConstraint Parser;
+            public readonly bool IsArray;
+            public readonly int Index;
+
+            public QueryParameterDestination(string componentParameterName, UrlValueConstraint parser, bool isArray, int index)
```

Comment:
FYI - an option is to define the fields as `init` properties, at which point you don't need to specify a ctor.

Reply:
I specifically avoided that because then you have properties, not fields. Since this is an internal struct there's no value in going through "getter" method calls for every field access.

CoreCLR might JIT that into something equivalent to field access (I don't know) but I'm definitely not assuming that `dotnet.wasm` interpreter would do so.